socs/agents/camera_rtsp/agent.py

```diff
@@ -0,0 +1,887 @@
+# Copyright (C) 2023-2024 Simons Observatory Collaboration
+# See top-level LICENSE.txt file for more information.
+"""Agent to capture images from cameras which support the RTSP protocol.
+"""
+
+import glob
+import os
+import re
+import time
+from collections import deque
+from datetime import datetime, timezone
+
+import cv2
+import imutils
+import numpy as np
+import ocs
+import txaio
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import Pacemaker, TimeoutLock
+
+txaio.use_twisted()
+
+
+def image_read_callback(path):
+    """Read callback for images.
+
+    Args:
+        path (str):  Path to the image file.
+
+    Returns:
+        (array):  Image data.
+
+    """
+    return cv2.imread(path)
+
+
+def image_write_callback(data, path, quality=20):
+    """Write callback for images.
+
+    Args:
+        data (array):  Image data.
+        path (str):  Path to the image file.
+        quality (int):  JPEG quality (0-100).
+
+    Returns:
+        None
+
+    """
+    cv2.imwrite(path, data, [int(cv2.IMWRITE_JPEG_QUALITY), quality])
+
+
+def video_read_callback(path):
+    """Read callback for video.
+
+    Args:
+        path (str):  Path to the video file.
+
+    Returns:
+        (array):  video data.
+
+    """
+    data = list()
+    cap = cv2.VideoCapture(path)
+    success = True
+    while success:
+        success, frame = cap.read()
+        if success:
+            data.append(frame)
+    cap.release()
+    return data
+
+
+def video_write_callback(data, path, fps=20.0):
+    """Write callback for video.
+
+    Args:
+        data (list):  List of frames.
+        path (str):  Path to the video file.
+
+    Returns:
+        None
+
+    """
+    first = data[0]
+    height = first.shape[0]
+    width = first.shape[1]
+    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
+    out = cv2.VideoWriter(path, fourcc, fps, (width, height))
+    for frame in data:
+        out.write(frame)
+    out.release()
+
+
+class CircularMediaBuffer:
```

Comment:
I'm not too familiar with circular media buffers, pardon my ignorance here: what happens when the agent crashes/exits and is restarted? Does this gracefully pick back up the buffer and continue handling files correctly?

Reply:
Yes, on construction it scans the existing files and picks up where it left off.  I'll move these helper classes into `socs/common/camera.py` for general use.